File: android/src/main/java/com/reactnativedeviceintel/RuntimeTimingProvider.kt
package com.reactnativedeviceintel

import android.os.SystemClock
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

internal class RuntimeTimingProvider {
  fun getRuntimeTimingSignals(): WritableMap {
    val intervals = mutableListOf<Double>()
    var previous = SystemClock.elapsedRealtimeNanos()
    repeat(SAMPLE_COUNT) {
      val current = SystemClock.elapsedRealtimeNanos()
      val delta = current - previous
      if (delta > 0) intervals.add(delta.toDouble())
      previous = current
    }
    val summary = SignalStatistics.summarize(intervals)
    return Arguments.createMap().apply {
      putString("nativeClockSource", "elapsed_realtime_nanos")
      putInt("nativeSampleCount", intervals.size)
      putDouble("nativeTimerResolutionNs", intervals.minOrNull() ?: 0.0)
      putDouble("nativeIntervalMedianNs", summary?.median ?: 0.0)
      putDouble("nativeIntervalP95Ns", summary?.p95 ?: 0.0)
      putDouble("nativeIntervalMadNs", summary?.mad ?: 0.0)
    }
  }

  private companion object {
    const val SAMPLE_COUNT = 256
  }
}

File: sdks/ios/Sources/IOSDeviceRiskSignals/RuntimeTimingProvider.m
#import "RuntimeTimingProvider.h"
#import "SignalStatistics.h"

static const NSUInteger kRuntimeTimingSamples = 256;

@implementation RuntimeTimingProvider

- (NSDictionary *)runtimeTimingSignals
{
  NSMutableArray<NSNumber *> *intervals = [NSMutableArray arrayWithCapacity:kRuntimeTimingSamples];
  CFTimeInterval previous = CFAbsoluteTimeGetCurrent();
  for (NSUInteger index = 0; index < kRuntimeTimingSamples; index++) {
    CFTimeInterval current = CFAbsoluteTimeGetCurrent();
    double deltaNs = (current - previous) * 1000000000.0;
    if (deltaNs > 0) [intervals addObject:@(deltaNs)];
    previous = current;
  }
  NSDictionary<NSString *, NSNumber *> *summary = RNDISummarize(intervals);
  NSNumber *resolution = [intervals valueForKeyPath:@"@min.self"] ?: @0;
  return @{
    @"nativeClockSource": @"cf_absolute_time",
    @"nativeSampleCount": @(intervals.count),
    @"nativeTimerResolutionNs": resolution,
    @"nativeIntervalMedianNs": summary[@"median"] ?: @0,
    @"nativeIntervalP95Ns": summary[@"p95"] ?: @0,
    @"nativeIntervalMadNs": summary[@"mad"] ?: @0,
  };
}

@end
